### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by zlalvani**
Preprocesses survey training data

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+python:
+  - "3.5"
+
+addons:
+  postgresql: "9.4"
+  redis: "3.2.5"
+
+before_script:
+  - psql -U postgres
+  - wget https://raw.githubusercontent.com/rpiml/okcollege-dev/master/assets/form.json -P ./assets
```
Comment (seveibar) [inline .travis.yml]: you should pull the entire assets directory :penguin:

File: docker-test.yml [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    network_mode: host
+    ports:
+      - "6379:6379"
+  survey-training-preprocessor-test:
+    image: python:3.5
+    network_mode: host
+    volumes:
+      - ../assets:/usr/src/app/assets
+      - ./:/usr/src/app
+    command: "/bin/bash -c \"cd /usr/src/app && pip install -r requirements.txt && python -u test.py\""
```
Comment (seveibar) [inline docker-test.yml:17]: what is python -u?

File: helpers.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
+import csv
+import json
+import pg8000
+import io
+
+def query_db(query, database='postgres', host='localhost', user='postgres', password='', response=True, autocommit=False):
+        psql_conn = pg8000.connect(host=host, database=database, user=user, password=password)
+        psql_conn.autocommit = autocommit
+
+
```
Comment (seveibar) [inline helpers.py:16]: extra spaces :penguin:

File: helpers.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
+import csv
+import json
+import pg8000
+import io
+
+def query_db(query, database='postgres', host='localhost', user='postgres', password='', response=True, autocommit=False):
```
Comment (seveibar) [inline helpers.py]: query_db should be pulled from environment variables

File: helpers.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+        psql_conn = pg8000.connect(host=host, database=database, user=user, password=password)
+        psql_conn.autocommit = autocommit
+
+
+        print('Querying DB')
+
+        cursor = psql_conn.cursor()
+        cursor.execute(query)
+
+        if response:
```
Comment (seveibar) [inline helpers.py]: needs comments

File: helpers.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+            result = cursor.fetchall()
+            cursor.close()
+            psql_conn.close()
+            return result
+        else:
+            cursor.close()
+            psql_conn.close()
+
+def process_survey_result(result, type_dict):
+    response_table = io.StringIO()
```
Comment (seveibar) [inline helpers.py:42]: needs comments,  doc string

File: helpers.py [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    writer = csv.writer(response_table, delimiter='\t')
+    for row in result:
+        try:
+            response_vector = []
+            seen = set()
```
Comment (seveibar) [inline helpers.py]: what is this

Comment (seveibar) [reply helpers.py]: "seen"

**Review by seveibar — CHANGES_REQUESTED** [review]

File: docker-test.yml (hunk 2 above)
Comment (zlalvani) [reply docker-test.yml:17]: flag for unbuffered output

File: test.py [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+            open('test_assets/example_response.json') as example_response:
+
+        helpers.query_db('CREATE TABLE survey_response (uuid uuid, content json)', database='postgres', response=False, autocommit=True)
+
+        ins_query = "INSERT INTO survey_response VALUES ('%s', '%s')" % \
+                    ("782786f0-a6fa-11e6-8a53-795ffb01e23b", json.dumps(json.load(example_response)))
+
+        helpers.query_db(ins_query, database='postgres', response=False, autocommit=True)
+
+        result = helpers.query_db('SELECT uuid, content FROM survey_response', database='postgres')
```
Comment (seveibar) [inline test.py:21]: redundant database specification, i think a singleton is appropriate here

File: test.py [hunk 9]
```diff
@@ -0,0 +1,50 @@
+import pg8000
+import json
+import redis
+import helpers
+
+if __name__ == '__main__':
```
Comment (seveibar) [inline test.py]: needs comments, would also recommend testing framework

File: test.py [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+        example_features_list = example_features.read().split('\n')
+        output_features_list = output_features.decode('utf-8').split('\r\n')
+
+        print('%d and %d should be identical' % (len(example_features_list), len(output_features_list)))
+        print(len(example_features_list), len(output_features_list))
+        assert(len(example_features_list) == len(output_features_list))
+
+        print('\nFollowing 2 tables should be identical:')
+        for i, line in enumerate(example_features_list):
+            print('%s <-> %s' % (line.strip(), output_features_list[i].strip()))
```
Comment (seveibar) [inline test.py]: this should all be broken up into functions and commented

File: preprocessor.py [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+            conn = amqpy.Connection(userid='rabbitmq', password='rabbitmq', host='localhost')
+            break
+        except Exception as e:
+            print("Could not connect to RabbitMQ. Retrying...")
+            time.sleep(1)
+
+    channel = conn.channel()
+    channel.exchange_declare('preprocessor', 'direct')
+    channel.queue_declare('survey-training-preprocessor')
+    channel.queue_bind('survey-training-preprocessor', exchange='preprocessor', routing_key='survey-training-preprocessor')
```
Comment (seveibar) [inline preprocessor.py:45]: I don't understand these, are they obvious? Why an exchange preprocesser? Why declare the exchange preprocessor/direct?

File: preprocessor.py [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+    print("Starting...")
+
+    print("Attempting connection...")
+    while True:
+        try:
+            conn = amqpy.Connection(userid='rabbitmq', password='rabbitmq', host='localhost')
+            break
+        except Exception as e:
+            print("Could not connect to RabbitMQ. Retrying...")
+            time.sleep(1)
```
Comment (seveibar) [inline preprocessor.py:40]: sleep for less time

**Review by seveibar — CHANGES_REQUESTED** [review]

File: helpers.py [hunk 13]
```diff
@@ -3,38 +3,54 @@
...
+    cursor.execute(query)
+
+    if response:
+        result = cursor.fetchall()
+        cursor.close()
+        psql_conn.close()
+        return result
+    else:
+        cursor.close()
+        psql_conn.close()
```
Comment (seveibar) [inline helpers.py:29]: possible change to singleton? Seems to weird to open and close connections every time :penguin:  should as an issue after merge

File: helpers.py [hunk 14]
```diff
@@ -3,38 +3,54 @@
...
 
 def process_survey_result(result, type_dict):
+    '''
+    This function takes:
+        - the result of a database query (as a list of JSON objects)
```
Comment (seveibar) [inline helpers.py:34]: label with appropriate parameter

Comment (seveibar) [reply helpers.py:34]: i.e. `result: the result of a database query (as a list of JSON objects)`

Comment (seveibar) [reply helpers.py:34]: :penguin:

**Review by seveibar — APPROVED** [review]
nice revision!

**Pull request merged**